core/src/main/java/org/apache/iceberg/io/ClusteredWriter.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.Set;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructLikeSet;
+
+/**
+ * A writer capable of writing to multiple specs and partitions ensuring the incoming records are properly clustered.
+ */
+public abstract class ClusteredWriter<T, R> implements PartitionAwareWriter<T, R> {
+
+  private final Set<Integer> completedSpecs = Sets.newHashSet();
+
+  private PartitionSpec currentSpec = null;
+  private Comparator<StructLike> partitionComparator = null;
+  private Set<StructLike> completedPartitions = null;
+  private StructLike currentPartition = null;
+  private Writer<T, R> currentWriter = null;
+
+  private boolean closed = false;
+
+  protected abstract Writer<T, R> newWriter(PartitionSpec spec, StructLike partition);
+
+  protected abstract void add(R result);
+
+  protected abstract R aggregatedResult();
+
+  @Override
+  public void write(T row, PartitionSpec spec, StructLike partition) throws IOException {
+    if (!spec.equals(currentSpec)) {
+      if (currentSpec != null) {
+        closeCurrent();
+        completedSpecs.add(currentSpec.specId());
+        completedPartitions.clear();
+      }
+
+      if (completedSpecs.contains(spec.specId())) {
+        throw new IllegalStateException("Already closed files for spec: " + spec.specId());
+      }
+
+      Types.StructType partitionType = spec.partitionType();
+
+      currentSpec = spec;
+      partitionComparator = Comparators.forType(partitionType);
+      completedPartitions = StructLikeSet.create(partitionType);
```

Comment:
Using `StructLikeSet` instead of a regular `Set` as we may get arbitrary `StructLike` implementations. This set is not going to be checked for every row so I don't worry a lot about performance here.

Reply:
I think the original relied on `PartitionKey` to handle hashing and so it was safe. That was before we introduced `StructLikeSet`, which is probably the easier way to go.

Reply:
Using `PartitionKey` is not even an option. For delete files, the partition struct will come from a metadata column, not by applying  `PartitionKey` on data. It seems `StructLikeSet` is a nice solution and since we are not calling it for every single row, it should not be much more expensive.